crates/bevy_utils/macros/src/lib.rs

```diff
@@ -35,12 +35,55 @@ impl Parse for AllTuples {
     }
 }
 
+/// Helper macro to generate tuple pyramids. Useful to generate scaffolding to work around Rust
+/// lacking variadics. Invoking `all_tuples!(impl_foo, start, end, P, Q, ..)`
+/// invokes `impl_foo` providing ident tuples through arity `start..=end`.
+/// ```
+/// use bevy_utils_proc_macros::all_tuples;
+///
+/// trait Foo {
+///     // ..
+/// }
+///
+/// macro_rules! impl_foo {
+///     ($($P:ident),*) => {
+///         // ..
+///     }
+/// }
+///
+/// all_tuples!(impl_foo, 0, 16, P);
+/// // impl_foo!();
+/// // impl_foo!(P0);
+/// // impl_foo!(P0, P1);
+/// // ..
+/// // impl_foo!(P0 .. P15);
+/// ```
+/// ```
+/// use bevy_utils_proc_macros::all_tuples;
+///
+/// trait Foo {
+///     // ..
+/// }
+///
+/// macro_rules! impl_foo {
+///     ($(($P:ident, $Q:ident)),*) => {
+///         // ..
```

Comment:
```suggestion
/// trait Foo<Q> {
///     // ..
/// }
///
/// macro_rules! impl_foo {
///     ($(($P:ident, $Q:ident)),*) => {
///         impl<$($P, $Q,)*> Foo<$($Q,)*> for ($($P,)*)
///         where $($P: Foo<$Q>),*
///         {
///             // ..
///         }
```

Same here, an example impl will make it more clear how this is supposed to be used (not sure if this is right, I haven't tried it in an IDE).